pkg/ocicni/ocicni.go

```diff
@@ -174,28 +183,113 @@ func (plugin *cniNetworkPlugin) Name() string {
 	return CNIPluginName
 }
 
-func (plugin *cniNetworkPlugin) SetUpPod(netnsPath string, namespace string, name string, id string) error {
+func (plugin *cniNetworkPlugin) monitorNetDir(netnsPath string, namespace string, name string, id string, timeout time.Duration) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logrus.Errorf("could not create new watcher", err)
+		return err
+	}
+	defer watcher.Close()
+
+	monitorTimeout := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				logrus.Debugf("CNI monitoring event %v", event)
+				if event.Op&fsnotify.Create != fsnotify.Create {
+					continue
+				}
+
+				if err := plugin.setUpPod(netnsPath, namespace, name, id); err != nil {
+					logrus.Errorf("could not set up pod with new CNI conf %v", err)
+				}
+
+				logrus.Debugf("CNI asynchronous setting succeeded")
+				plugin.terminateMonitorNetDir()
+				return
+
+			case err := <-watcher.Errors:
+				logrus.Errorf("CNI monitoring error %v", err)
+				plugin.terminateMonitorNetDir()
+				return
+
+			case <-monitorTimeout:
+				logrus.Errorf("CNI monitoring timeout")
+				return
+			}
+		}
+	}()
+
+	if err = watcher.Add(plugin.pluginDir); err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	select {
+	case <-plugin.monitorNetDirChan:
+		logrus.Debugf("CNI monitoring finished")
+
+	case <-time.After(time.Second * timeout):
+		close(monitorTimeout)
+	}
+
+	return nil
+}
+
+func (plugin *cniNetworkPlugin) setUpPod(netnsPath string, namespace string, name string, id string) error {
+	plugin.syncNetworkConfig()
+
 	if err := plugin.checkInitialized(); err != nil {
```

Comment:
@sameo Do we need to run this again here? The parent SetUpPod already called it once over.